<h2> Задача </h2>

Известно, что вклад, находящийся в банке с начала года, возрастает к концу года на определенный процент (свой для каждого банка). В начале года 3/8 вклада, который составляет 800 тыс. руб., вложили в первый банк, 1/8 во второй банк и оставшуюся часть вклада в третий банк. К концу года сумма этих вкладов стала равна 907 тыс. руб. Если бы первоначально 1/8 вклада положили в первый банк, 4/8 вклада — во второй банк, оставшуюся часть вклада — в третий банк, то к концу года сумма этих вкладов стала бы равна 894 тыс. руб. Если бы 4/8 вклада вложили в первый банк, 3/8 вклада — во второй банк, оставшуюся часть вклада — в третий банк, то к концу года сумма этих вкладов была бы равна 903 тыс. руб. Какой процент начисляет каждый банк?

<h3> Решение: </h3>

Введем обозначения: <br>
Пусть $x_{1}$ - процент, начисляемый вкладчику в первом банке, тогда $x_{2}, x_{3}$ проценты во втором и третьем банке соответственно. Расчитаем вклады, умножив части вклада на общую его сумму: <br>
$$
\frac{3}{8}\cdot 800 = 300 ,
\frac{1}{8}\cdot 800 = 100 ,
\frac{4}{8}\cdot 800 = 400
$$


In [1]:
cost = 800 # тыс.руб
cost1 = 0.375 * cost
cost2 = 0.125 * cost
cost3 = 0.5 * cost
print(f"Суммы вкладов в банки 1, 2 и 3: {cost1} тыс.руб, {cost2} тыс.руб, {cost3} тыс.руб.")

Суммы вкладов в банки 1, 2 и 3: 300.0 тыс.руб, 100.0 тыс.руб, 400.0 тыс.руб.


Расчитаем сумму с начислеными процентами за год в каждом из банков по формуле: $cost_{i}\cdot \frac{x_{1}}{100} $ <br>
Начислено в первом банке: $300\cdot \frac{x_{1}}{100} = 3x_{1}$ <br>
Начислено во втором банке: $ 100\cdot \frac{x_{1}}{100} = x_{2} $ <br>
Начислено в третьем банке: $ 400\cdot \frac{x_{1}}{100} = 4x_{3} $ <br>

Всего на вклад в 800 тыс. руб., сделанный в три банка (в первый — 300 тыс. руб., во второй — 100 тыс. руб., в третий — 400 тыс. руб.), было начислено за год: 907 — 800 = 107 тыс. руб. Следовательно, можем составить систему уравнений. <br>
Первое уравнение системы: $ 3x_{1}+x_{2}+4x_{3} = 107 $ <br>
Следующие уравнения составляем аналогично: <br>
$$
x_{1}+4x_{2}+3x_{3}=94, 4x_{1}+3x_{2}+x_{3}=103
$$
Получим систему трех линейных уравнений:
$$
\left\{\begin{matrix}
3x_{1} + x_{2} + 4x_{3} = 107 \\ 
x_{1} + 4x_{2} + 3x_{3} = 94 \\
4x_{1} + 3x_{2} + x_{3} = 103
\end{matrix}\right.
$$



Запишем уравнение в виде матрицы A и матрицы столбца свободных членов B:
$$
\left(\begin{array}{ccc|c}  
 3 & 1 & 4 & 107 \\  
 1 & 4 & 3 & 94\\ 
 4 & 3 & 1 & 103\\ 
\end{array}\right)
$$


Решим при помощи библиотеки numpy:

In [2]:
import numpy as np
A = np.array([[3., 1., 4.],[1., 4., 3.], [4., 3., 1.]]) # Коэффициенты
B = np.array([107., 94., 103.]) # Свободные члены
result = np.linalg.solve(A, B)
print(f"Решение системы библиотечным методом: {result}")


Решение системы библиотечным методом: [15. 10. 13.]


Решение методом Гаусса, проверка результата:

In [3]:
def gaussian(A, b):
    # Расширенная матрица
    reshaped_b = b.reshape((len(b), 1))
    A = np.hstack((A, reshaped_b))
    
    # Приведение матрицы к треугольному виду
    for i, i_val in enumerate(A):
        for j in range(i + 1, len(A)):
            A[j] -= A[i] * A[j][i] / A[i][i]
            
    # Подстановка (обратный ход)
    x = np.array([0] * len(b), dtype=float)
    i = len(A) - 1
    while i >= 0:
        x[i] = (A[i][-1] - sum(x * A[i][0:-1])) / A[i][i] 
        i -= 1
    return x
    
print("Решение системы методом Гаусса: ", gaussian(A, B))

Решение системы методом Гаусса:  [15. 10. 13.]


Система имеет единственное решение: $x_{1} = 15, x_{2} = 10, x_{3} = 13$ <br>

<b> Ответ: </b> первый банк выплачивает 15% годовых, второй банк - 10% годовых, а третий банк - 13% годовых.
